base/linalg/dense.jl

```diff
@@ -309,17 +309,60 @@ kron(a::AbstractVector, b::AbstractVector)=vec(kron(reshape(a,length(a),1),resha
 kron(a::AbstractMatrix, b::AbstractVector)=kron(a,reshape(b,length(b),1))
 kron(a::AbstractVector, b::AbstractMatrix)=kron(reshape(a,length(a),1),b)
 
-^(A::AbstractMatrix, p::Integer) = p < 0 ? inv(A^-p) : Base.power_by_squaring(A,p)
+# Matrix power
+function ^{T}(A::AbstractMatrix{T}, p::Real)
+    # non-inversible matrix to negative power is undefined.
+    if (p < 0) && (det(A) == 0)
+        throw(ArgumentError("Negative powers of singular matrix are not defined"))
+    end
 
-function ^(A::AbstractMatrix, p::Number)
+    # For integer powers, use repeated squaring
     if isinteger(p)
-        return A^Integer(real(p))
+        return p < 0 ? Base.power_by_squaring(inv(A),-Integer(p)) : Base.power_by_squaring(A,Integer(p))
```

Comment:
If `A` is an integer matrix, `A^2` should be an integer matrix, but `A^2.0` should be a floating-point matrix.

We definitely should *not* use diagonalization for integer `p`, because it will be way slower even than what we are doing now, despite the temporary matrices.  e.g.
```jl
julia> A = randn(1000,1000); A = A'*A;

julia> @time A^4;
  0.428707 seconds (114.28 k allocations: 20.806 MiB, 20.50% gc time)

julia> @time A^4.5;
  3.669913 seconds (2.70 M allocations: 172.916 MiB, 2.36% gc time)
```